### ISSUE TITLE
Error setting up local client in OS X & Linux, FileNotFoundError: [Errno 2] No such file or directory: '7zr' 

### ISSUE THREAD

**Issue by banagale** [s3inlc/hashtopolis#387]
I'm trying run the hashtopolis client on os x as a test agent for my vps-hosted copy of hashtopolis. 

- Server version: 0.5.1
- Client version: 's3-python-0.1.3'
- Hashcat version: v4.1.0-29-g547025ec

I'm trying to run `python hashtopolis.zip`

The output I get is:

```
Starting client 's3-python-0.1.3'...
Collecting agent data...
Login successful!
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "hashtopolis.zip/__main__.py", line 124, in <module>
  File "hashtopolis.zip/__main__.py", line 42, in init
  File "hashtopolis.zip/htpclient/binarydownload.py", line 19, in run
  File "hashtopolis.zip/htpclient/binarydownload.py", line 41, in __check_utils
FileNotFoundError: [Errno 2] No such file or directory: '7zr'
```

I figured this was trying to download the hashcat binary I added on the server. Since it was failing, I tried building hashcat directly and placing that in the /crackers/ folder.

I was able to execute `./hashcat -b -m 2500`, however, `python hashtopolis.zip` continues to attempt to download the a binary and seemingly fails to find it.

I do see on my server Agents list that my client machine was found and its last activity was `downloadBinary at`

Edit: I attempted the same setup routine in on the linux box the server is running on and encountered the same error,

**Comment by s3inlc** [comment]
The error does not happen when downloading the hashcat binary, it happens earlier when it tries to get the 7z binary for extraction.
Can you please check that on the server there are three 7zr files in the src/static/ folder and also if on the client there was any file downloaded into the client directory which is named 7zr.
The exception itself happens when it tries to apply the execution flag the binary. Is it possible that this is blocked by some software?

**Comment by banagale** [comment]
Thank you for the reply. The server did have three 7zr files, however the client directory did not have a folder named 7zr. I simply created this folder in the client's main directory and it appears the client is now running as expected.

I also manually added the `7zr` directory to the root of the hashtopolis client i put on the same box as the server and now that appears to be running as well.

My assumption is `No task available!` and the last activity on the agents list as getTask is that things are running as expeceted.

**Comment by banagale** [comment]
I ran into some additional errors, fixed them, I thought and then am stuck again.

I added a word list, and pasted in a small md5 hash list. I created a task and assigned it to my mac client first.

The client picked it up right away but then errored out:

```
Got task with id: 1
Downloading: [==================================================]
sh: ./7zr: is a directory
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "hashtopolis.zip/__main__.py", line 125, in <module>
  File "hashtopolis.zip/__main__.py", line 66, in loop
  File "hashtopolis.zip/htpclient/binarydownload.py", line 71, in check_version
FileNotFoundError: [Errno 2] No such file or directory: 'crackers/temp'
```

I created the directory crackers/temp and ran the client again and it errored out:

```
(hashtop) ➜  hashtopolis python hashtopolis.zip 
Starting client 's3-python-0.1.3'...
Collecting agent data...
Login successful!
Got task with id: 1
Downloading: [==================================================]
sh: ./7zr: is a directory
Downloading: [==================================================]
Got cracker binary type hashcat
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "hashtopolis.zip/__main__.py", line 125, in <module>
  File "hashtopolis.zip/__main__.py", line 88, in loop
  File "hashtopolis.zip/htpclient/hashcat_cracker.py", line 189, in measure_keyspace
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 336, in check_output
    **kwargs).stdout
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 403, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 707, in __init__
    restore_signals, start_new_session)
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 1326, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'crackers/1/'
```

I created `crackers/1` and ran it again:

```
(hashtop) ➜  hashtopolis python hashtopolis.zip
Starting client 's3-python-0.1.3'...
Collecting agent data...
Login successful!
Got task with id: 1
Downloading: [==================================================]
Got cracker binary type hashcat
/bin/sh: ./hashcat-4.1.0.7z64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat-4.1.0.7z64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat-4.1.0.7z64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat-4.1.0.7z64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat-4.1.0.7z64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat-4.1.0.7z64.osx: No such file or directory
```

I created `crackers/1/hashcat-4.1.0.7z64.bin/` and ran it again and it now runs but says `No task available. 

I went through the same process w the linux client.

The created directory for hashcat seems to empty. I wonder if the previous error put the client in a bad state because no matter the job I create now, neither client seems to see it. (I can assign but nothing happens client side)

I have the compiled hashcat binary available, is there a way I can manually put it in? Any ideas on how to get the client to start accepting jobs again?

**Comment by s3inlc** [comment]
7zr should not be a directory. It should be the appropriate 7z binary from the server (depending on the OS of the agent). As long as it does not has the right 7zr binary, it is not able to extract hashcat after downloading.
To fix it manually you should copy the 7zr.osx from the server into the client folder and name it `7zr`. Also what I see from your debug output, your cracker config is not correctly, you should not put the full name of the hashcat 7z as cracker name. There should only be `hashcat`.

**Comment by banagale** [comment]
Thanks for the additional notes. I renamed the base file as it should have been, and I copied over 7zr.osx as simply `7zr` to the hashtopolis base path. Downloading directly from [here](https://github.com/s3inlc/hashtopolis/blob/master/src/static/7zr.osx). This allowed the agent to run to repeated `No task available!` 

I was able to create a new task, and assign it to the agent, (after marking it okay for secret data) and it took up the task but errored out. I realized I also needed to set executable permissions on 7zr. 

After doing that and running again, I got to here:

```
(hashtop) ➜  hashtopolis python hashtopolis.zip
Starting client 's3-python-0.1.3'...
Collecting agent data...
Login successful!
Got task with id: 3
Downloading: [==================================================]

7-Zip (a) [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,4 CPUs x64)

Scanning the drive for archives:
1 file, 2742468 bytes (2679 KiB)

Extracting archive: crackers/1.7z
--
Path = crackers/1.7z
Type = 7z
Physical Size = 2742468
Headers Size = 8357
Method = LZMA2:25 LZMA:20 BCJ2
Solid = +
Blocks = 2

Everything is Ok                                      

Folders: 41
Files: 909
Size:       26483880
Compressed: 2742468
Downloading: [==================================================]
Downloading: [========================================================================================]
Got cracker binary type hashcat
/bin/sh: ./hashcat64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat64.osx: No such file or directory
Error during keyspace measure
/bin/sh: ./hashcat64.osx: No such file or directory
Error during keyspace measure
```

**Comment by banagale** [comment]
I went ahead and compiled `hashcat` for os x and put it in the `crackers/1/` folder renamed as `hashcat64.osx` and set up another plan and ran it against it again. It worked this time! I will try again on the linux client.

**Comment by s3inlc** [comment]
Based on your first descriptions on your issue, there must have been a problem downloading the 7zr from the server (as it should do this automatically) and then ran into further problems because of missing this.
Glad to see that it seems to work now for you. OS X always is a bit special as there are not pre-builds from hashcat.net available. But when you want to distribute it automatically (instead of compiling it on every agent and put it into `crackers/1`, you can build it once and then pack it into the normal hashcat 7z file, provide it somewhere to download which you then can set on the Cracker settings.

**Comment by winxp5421** [comment]
Closing this issue due to inactivity. If you have any other issues please reopen or create a new issue ticket
